addons/web/controllers/json.py

```diff
@@ -0,0 +1,259 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import ast
+import functools
+import logging
+from collections import defaultdict
+from datetime import date
+from http import HTTPStatus
+from urllib.parse import urlencode
+
+import psycopg2.errors
+from dateutil.relativedelta import relativedelta
+from lxml import etree
+from werkzeug.exceptions import BadRequest
+
+from odoo import http
+from odoo.exceptions import AccessError
+from odoo.http import request
+from odoo.models import regex_object_name
+from odoo.osv import expression
+from odoo.tools import lazy
+from odoo.tools.safe_eval import safe_eval
+
+from .utils import get_action_triples
+
+_logger = logging.getLogger(__name__)
+
+
+class WebJsonController(http.Controller):
+
+    @http.route('/json/<path:subpath>', auth='public', type='http', readonly=True)
+    def web_json(self, subpath, **kwargs):
+        return self.web_json_version_1(subpath, **kwargs, redirect=True)
+
+    @http.route('/json/1/<path:subpath>', auth='user', type='http', readonly=True)
+    def web_json_version_1(self, subpath, **kwargs):
+        env = request.env
+        if not request.env.user.has_group('base.group_allow_export'):
+            raise AccessError(env._("You need export permissions to use the /json route"))
+
+        # redirect when the kwargs have changed
+        param_list = set(kwargs)
+        kwargs.pop("redirect", None)  # pop redirect, so it is forced in check_redirect
+
+        def check_redirect():
+            if set(param_list) == set(kwargs):
+                return None
+            # for domains, make chars as safe
+            encoded_kwargs = urlencode(kwargs, safe="()[], '\"")
+            return request.redirect(
+                f'/json/1/{subpath}?{encoded_kwargs}',
+                HTTPStatus.TEMPORARY_REDIRECT
+            )
```

Comment:
The goal is to redirect the user to a URL that contains the parameters explictely.

For example: `/json/crm?view_type=list` becomes `/json?crm?view_type=list&domain=[]&offset=0&limit=80`.

Reply:
While a good idea when navigating interactively, not sure it is wise for scripts. I double people will set all those parameters all the time. I think they're going to blindly rely on their library to follow the redirection, every time, doubling the requests and round-trip time. What's about setting the [`Content-Location`](https://httpwg.org/specs/rfc9110.html#field.content-location) header instead?

Reply:
> recipient MAY consider the content to be a current representation of that resource at the time indicated by the message origination date.

While the browser could cache the result, it does not update the URL (needed feature). For this, I don't know any other possibility than a redirect.

Actually, the main use case for /json/1 is exploration: the user would copy-paste the URL and adapt the parameters before putting it into a scrript - because all parameters are set, there would be no redirect.